hermes-frontend/src/main/java/pl/allegro/tech/hermes/frontend/publishing/handlers/DynamicThroughputLimiter.java

```diff
@@ -0,0 +1,101 @@
+package pl.allegro.tech.hermes.frontend.publishing.handlers;
+
+import com.codahale.metrics.Metered;
+import pl.allegro.tech.hermes.api.TopicName;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+public class DynamicThroughputLimiter implements ThroughputLimiter, Runnable {
+    private final long max;
+    private final long threshold;
+    private final long desired;
+    private final double idleThreshold;
+    private final Metered globalThroughputMeter;
+
+    private final ScheduledExecutorService executor;
+    private final int checkInterval;
+
+    private ConcurrentHashMap<TopicName, Throughput> users = new ConcurrentHashMap<>();
+
+    public DynamicThroughputLimiter(long max,
+                                    long threshold,
+                                    long desired,
+                                    double idleThreshold,
+                                    int checkInterval,
+                                    Metered globalThroughput,
+                                    ScheduledExecutorService executor) {
+        this.max = max;
+        this.threshold = threshold;
+        this.desired = desired;
+        this.idleThreshold = idleThreshold;
+        this.globalThroughputMeter = globalThroughput;
+        this.checkInterval = checkInterval;
+        this.executor = executor;
+    }
+
+    @Override
+    public void check(TopicName topic, Metered rate) {
```

Comment:
this method uses exceptions as flow control - i think it should return proper values instead.